transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java

```diff
@@ -389,13 +393,47 @@ protected boolean offerChannel(Channel channel) {
 
     @Override
     public void close() {
+        closeIdleChannels(nextEventLoop());
+    }
+
+    void closeIdleChannels(EventExecutor executor) {
+        Promise<Void> result = executor.newPromise();
+        closeIdleChannels(result);
+        if (!executor.inEventLoop()) {
+            result.syncUninterruptibly();
+        }
+    }
+
+    private void closeIdleChannels(final Promise<Void> result) {
+        final Set<Channel> channelsToClose = Collections.newSetFromMap(new ConcurrentHashMap<Channel, Boolean>());
         for (;;) {
             Channel channel = pollChannel();
             if (channel == null) {
                 break;
             }
-            // Just ignore any errors that are reported back from close().
-            channel.close().awaitUninterruptibly();
+            channelsToClose.add(channel);
+        }
+        if (channelsToClose.isEmpty()) {
+            // no idle channels in the pool - nothing to close
+            result.trySuccess(null);
+        }
+
+        for (final Channel channel : channelsToClose) {
+            channel.close().addListener(new ChannelFutureListener() {
+                @Override
+                public void operationComplete(ChannelFuture future) {
+                    channelsToClose.remove(channel);
```

Comment:
Could you please elaborate on why is this a bad idea? Set of channels is thread-safe and channels are removed from it by event loop threads that execute close future callbacks. It could happen that channel is removed by the current thread but this should not cause any problems.

I'm not sure how CountDownLatch can be used here. It only allows code to block waiting for the count to be zero. Blocking is not desirable here, I believe.

It could be an `AtomicInteger` for counting closed channels. However, I feel like the current code is a bit more explicit about what's going on. Set of channels to be closed is "snapshotted" and every closed channel is removed from the set.